sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -153,6 +154,368 @@ object AggregateModeInfo {
+          withRetryAndMerge(preProcessed, splitPolicy, mergePolicy) { preProcessedAttempt =>
+            aggregate(preProcessedAttempt)
+          }
+        Seq(merged)
+      } else {
+        withRetry(preProcessed, splitPolicy) { preProcessedAttempt =>
+          aggregate(preProcessedAttempt)
+        }
+      }
+    }
+
+    /**
+     * Invoke reduction functions as defined in each `CudfAggreagte`
+     * @param preProcessed - a batch after the "pre" step
+     * @return
+     */
+    def performReduction(preProcessed: ColumnarBatch): ColumnarBatch = {
+      withResource(new NvtxRange("reduce", NvtxColor.BLUE)) { _ =>
+        val cvs = mutable.ArrayBuffer[GpuColumnVector]()
+        cudfAggregates.zipWithIndex.foreach { case (cudfAgg, ix) =>
+          val aggFn = cudfAgg.reductionAggregate
+          val cols = GpuColumnVector.extractColumns(preProcessed)
+          val reductionCol = cols(aggOrdinals(ix))
+          withResource(aggFn(reductionCol.getBase)) { res =>
+            cvs += GpuColumnVector.from(
+              cudf.ColumnVector.fromScalar(res, 1), cudfAgg.dataType)
+          }
+        }
+        new ColumnarBatch(cvs.toArray, 1)
+      }
+    }
+
+    /**
+     * Used to produce a group-by aggregate
+     * @param preProcessed the batch after the "pre" step
+     * @return a Table that has been cuDF aggregated
+     */
+    def performGroupByAggregation(preProcessed: ColumnarBatch): ColumnarBatch = {
+      withResource(new NvtxRange("groupby", NvtxColor.BLUE)) { _ =>
+        withResource(GpuColumnVector.from(preProcessed)) { preProcessedTbl =>
+          val groupOptions = cudf.GroupByOptions.builder()
+              .withIgnoreNullKeys(false)
+              .withKeysSorted(isSorted)
+              .build()
+
+          val cudfAggsOnColumn = cudfAggregates.zip(aggOrdinals).map {
+            case (cudfAgg, ord) => cudfAgg.groupByAggregate.onColumn(ord)
+          }
+
+          // perform the aggregate
+          val aggTbl = preProcessedTbl
+              .groupBy(groupOptions, groupingOrdinals:_*)
+              .aggregate(cudfAggsOnColumn: _*)
+
+          withResource(aggTbl) { _ =>
+            GpuColumnVector.from(aggTbl, postStepDataTypes.toArray)
+          }
+        }
+      }
+    }
+
+    /**
+     * Used to produce the outbound batch from the aggregate that could be
+     * shuffled or could be passed through the evaluateExpression if we are in the final
+     * stage.
+     * It takes a cuDF aggregated batch and applies the "post" step:
+     * postUpdate for update, or postMerge for merge
+     * @param resultBatch - cuDF aggregated batch
+     * @return output batch from the aggregate
+     */
+    def postProcess(resultBatch: ColumnarBatch): ColumnarBatch = {
+      withResource(new NvtxRange("post-process", NvtxColor.ORANGE)) { _ =>
+        GpuProjectExec.project(resultBatch, postStepBound)
+      }
+    }
+  }
+
+  lazy val splitPolicy: SpillableColumnarBatch => Seq[SpillableColumnarBatch] = {
```

Comment:
It's a bit weird for this to be a field of an instance rather than a static method, as it doesn't seem that this needs any instance state.  In looks like this just generically splits a batch in half, and as such should be a utility method that anyone can use rather than specific to aggregate.

Reply:
Yes, this is the generic I'll split a batch in half using the row counts. I'll move to a utility as it should be reusable. 

Writing this I wonder if we need one for byte counts (e.g. approximately break this chunk into two of similar byte sizes). That said, we can cross that bridge when needed. For example, `GpuGenerate` falls under this category, there we want to split based on what _would be_ the final byte size.